### PR TITLE
[fix] Retry a lost storage-unit request instead of failing the job

### PR DESCRIPTION
**Pull request by adoda**
## Problem

A storage unit stops answering and every client routed to it fails when its own recv timeout
expires. Raising the timeout (400s to 1800s in our deployment) changed nothing.

Evidence from one such failure: the unit's node was alive and serving other traffic throughout, a
TCP connect to its `put_get_socket` succeeded, and the unit's own counters showed it fully healthy
(1025 GET_DATA served, 9.6ms p99, 1.33GB RSS). It had served exactly one GET_DATA fewer than its
cohort. So the unit never saw the request that timed out; it was lost between the two ends, not
queued behind slow work. #168 protected the worker thread from dying, which is a different cause of
the same symptom; here the thread was intact.

That loss is invisible by construction. ZMQ connect is asynchronous and SNDHWM is 0, so a DEALER
accepts `send()` into an unbounded local queue for a peer it has not reached yet. The caller only
learns anything when its own RCVTIMEO expires, which is why no timeout value can distinguish a lost
request from a slow one. Lowering SNDHWM would not help either, it only trades silent queuing for
silent dropping.

## Fix

1. **Retry on a new connection.** Up to `TQ_SIMPLE_STORAGE_MAX_ATTEMPTS` (default 3, floored at 1).
   A fresh socket and TCP connection is the part that matters. Only a missing answer is retried:
   `zmq.error.Again` now raises `StorageUnitTimeout`, while an error the unit reported surfaces on
   the first attempt. Replaying is safe where it is used: get is read-only and put overwrites the
   same global indexes. A put carrying a `data_parser` is the exception and is sent once, because a
   retry would re-run the parser on the unit and the public API does not constrain its side
   effects.
2. **Self-diagnosing residual failure.** After the last attempt, ask the unit for its own counters
   over a fresh socket with a short timeout. That probe is served by the same worker thread as put
   and get, so an answer proves the request was lost in flight and silence means the unit itself
   stopped. The failure log carries that verdict plus tcp reachability, the unit's op counts and RSS,
   and the wire size of the request that failed (measured from the serialized ZMQ frames on put).
3. **Log volume unchanged in the steady state.** A recovered request logs one line and skips the
   diagnosis entirely. Heavy requests are logged only above `TQ_STORAGE_SLOW_REQUEST_SECONDS` (5s)
   or `TQ_STORAGE_LARGE_PAYLOAD_MB` (256MB), via shared `log_heavy_operation`: puts on the manager
   (serialized request frames + end-to-end RTT), gets on the unit (serialized response frames +
   local handling time). No tensor-`nbytes` estimate remains.

## Tests

`tests/test_storage_request_retry.py`: 14 passed. Covers recovery on retry, the bounded attempt
count, that a nonpositive count still issues one request, that errors reported by the unit are not
retried, that a parser-backed put is not replayed while the ordinary put still is, that retry logs
carry wire-size detail from the failed attempt, shared heavy-request logging thresholds, and each
diagnosis verdict.

Full non-e2e run: 623 passed, 10 skipped, plus `python -m compileall -q transfer_queue tutorial tests`.
`test_yuanrong_storage_client_e2e.py` is excluded locally, it needs the optional
`openyuanrong-datasystem` dependency.

### CODE REVIEW

**Comment by adoda** [comment]
> /check-cla

File: transfer_queue/storage/managers/simple_storage_manager.py [hunk 1]
```diff
@@ -257,24 +381,37 @@ async def put_data(
...
+                    "put",
+                    su_id,
+                    f"samples={len(group.global_indexes)} fields={list(storage_data.keys())} "
+                    f"payload_mb={estimate_payload_bytes(storage_data) / 2**20:.1f}",
+                    partial(
+                        self._put_to_single_storage_unit,
+                        group.global_indexes,
+                        storage_data,
+                        target_storage_unit=su_id,
+                        data_parser=data_parser,
```
[the remark this answers is not in the document]
Comment (adoda) [reply transfer_queue/storage/managers/simple_storage_manager.py:441]: Confirmed and fixed in 613bba8. `kv_put` / `kv_batch_put` / `put` document a contract for `data_parser` covering keys, element count and ordering, but never side effects, so treating replay-safety as an unwritten requirement was wrong.

Rather than dedupe at the unit, which needs per-request state there, a parser-backed put is now sent once: `put_data` passes `max_attempts=1` when `data_parser is not None`, so it fails exactly as it did before this change. The retry stays in force for the ordinary put, which is a plain overwrite of the same global indexes.

Tests: `test_parser_backed_put_is_not_replayed` asserts one attempt, `test_put_without_a_parser_is_still_retried` asserts the retry is not lost for the ordinary path.

File: transfer_queue/storage/managers/simple_storage_manager.py [hunk 2]
```diff
@@ -144,6 +171,103 @@ def _group_by_hash(self, global_indexes: list[int]) -> dict[str, RoutingGroup]:
...
+            operation: Operation name for logs, e.g. ``get`` or ``put``.
+            target_storage_unit: Unit this request is routed to.
+            request_context: Request shape, so both ends of a failure can be correlated.
+            make_request: Zero-arg callable returning a coroutine for one attempt. Must build a
+                new socket per call, which ``with_storage_unit_socket`` does.
+        """
+        endpoint = self._describe_storage_unit(target_storage_unit)
+        for attempt in range(1, TQ_SIMPLE_STORAGE_MAX_ATTEMPTS + 1):
+            try:
+                return await make_request()
```
[the remark this answers is not in the document]
Comment (adoda) [reply transfer_queue/storage/managers/simple_storage_manager.py]: Confirmed and fixed in 177667f. Reproduced first: with the count patched to 0, `_request_with_retry` returned `None` without raising, and for a put that reads as success, after which `put_data` would notify the controller with metadata for data that was never sent.

The attempt count is now floored at one where it is consumed, `attempts_allowed = max(1, ...)`, so every caller gets one attempt regardless of how the value was configured. I put the floor at the use site rather than at import so it also covers the internal `max_attempts` override, and so it is testable without reloading the module.

Test: `test_a_nonpositive_attempt_count_still_issues_one_request`.

File: transfer_queue/storage/managers/simple_storage_manager.py [hunk 3]
```diff
@@ -257,24 +387,40 @@ async def put_data(
         field_schema = extract_field_schema(data)
 
         routing = self._group_by_hash(metadata.global_indexes)
-        tasks = [
-            self._put_to_single_storage_unit(
-                group.global_indexes,
-                {f: self._select_by_positions(data[f], group.batch_positions) for f in data.keys()},
-                target_storage_unit=su_id,
-                data_parser=data_parser,
+        # Replaying a put is safe because it overwrites the same global indexes, but a retry also
```
Comment (0oshowero0) [inline transfer_queue/storage/managers/simple_storage_manager.py]: We can remove some comments

File: transfer_queue/utils/common.py [hunk 4]
```diff
@@ -133,3 +135,22 @@ def get_env_bool(env_key: str, default: bool = False) -> bool:
 
     true_values = {"true", "1", "yes", "y", "on"}
     return env_value_lower in true_values
+
+
+def estimate_payload_bytes(field_data: Any) -> int:
```
Comment (0oshowero0) [inline transfer_queue/utils/common.py]: This estimation might be highly inaccurate. Maybe a better solution is let the underlying `_put_to_single_storage_unit` to return the payload size and throw an error that will be captured by the retry layer?

For instance, to get the accurate size, we can
```python
frames = request_msg.serialize()
serialized_bytes = sum(memoryview(frame).nbytes for frame in frames)

await socket.send_multipart(frames, copy=False)
```
And raise when meeting errors to let retry layer know:
```python
except zmq.error.Again as exc:
    serialized_bytes = serialized_nbytes(frames)
    raise StorageUnitTimeout(
        f"put timeout: unit={target_storage_unit}, "
        f"samples={len(global_indexes)}, "
        f"serialized_bytes={serialized_bytes}"
    ) from exc
```

File: transfer_queue/storage/simple_storage.py [hunk 5]
```diff
@@ -394,6 +399,17 @@ def _worker_routine(self) -> None:
         poller.unregister(worker_socket)
         worker_socket.close(linger=0)
 
+    def _log_if_heavy(self, operation: str, started: float, field_data: Any, num_samples: int, fields: Any) -> None:
```
Comment (0oshowero0) [inline transfer_queue/storage/simple_storage.py]: Now the log only happens at simple storage side. Maybe a better choice is log for `put` in manager side, and log for `get` in simple storage side?

File: transfer_queue/storage/managers/simple_storage_manager.py (hunk 3 above)
Comment (adoda) [reply transfer_queue/storage/managers/simple_storage_manager.py]: Done in 44b7573: shortened to one line that states the constraint the `max_attempts=1` branch is enforcing.

File: transfer_queue/utils/common.py (hunk 4 above)
Comment (adoda) [reply transfer_queue/utils/common.py]: Agreed and fixed in 2bcbcf8. `estimate_payload_bytes` only walked tensor `nbytes`, so NonTensorStack / object fields, msgpack/pickle overhead, and multipart framing were all invisible, and a `data_parser` that expands references was measured before expansion.

`_put_to_single_storage_unit` now measures the frames after `serialize()` with `frame_nbytes`, puts `serialized_mb` on `StorageUnitTimeout`, and the retry layer logs the exception text so the wire size shows up on both intermediate retries and the final failure. The call-site `request_context` no longer estimates payload size.

File: transfer_queue/storage/simple_storage.py (hunk 5 above)
Comment (adoda) [reply transfer_queue/storage/simple_storage.py]: Done in 91a9e46. Put heavy logging moved to the manager, using the same serialized frame size and the end-to-end RTT around send/recv. Get heavy logging stays on the unit, where the response payload is built. The unit-side PUT `_log_if_heavy` call is gone.

File: transfer_queue/storage/managers/simple_storage_manager.py [hunk 6]
```diff
@@ -286,6 +435,20 @@ async def put_data(
             field_schema,
         )
 
+    def _log_if_heavy_put(
```
Comment (0oshowero0) [inline transfer_queue/storage/managers/simple_storage_manager.py]: We better extract this as a util function `log_heavy_operation`, passing id, criteria, warning message into it. Therefore we can reuse it for `SimpleStorage` side

File: transfer_queue/utils/common.py (hunk 4 above)
Comment (0oshowero0) [reply transfer_queue/utils/common.py]: We can also refactor the get side accordingly, and totally remove this function

File: transfer_queue/storage/managers/simple_storage_manager.py (hunk 6 above)
Comment (adoda) [reply transfer_queue/storage/managers/simple_storage_manager.py]: Done in f3d31af. Extracted `log_heavy_operation(component_id, operation, elapsed, payload_bytes, detail)` into `transfer_queue/utils/common.py` next to the shared thresholds, and both the manager put path and the unit get path call it.

File: transfer_queue/utils/common.py (hunk 4 above)
Comment (adoda) [reply transfer_queue/utils/common.py]: Done in the same commit. Get now measures serialized response frames the same way put measures request frames, so `estimate_payload_bytes` has no callers left and is removed.

**Comment by adoda** [comment]
Pushed 7f28a13. The retry logs repeated what the exception already said: the unit id appeared three times in one line, the endpoint, `samples=` and the timeout twice each. `StorageUnitTimeout` now owns unit + endpoint + timeout (that message is what `put_data` / `get_data` callers see), and the retry lines add only the attempt and the request shape. Put and get now report the same fields; get previously carried no request detail at all while put carried two.

```
[TQ_STORAGE_9f2a]: put failed after 3 attempts. samples=64 fields=['input_ids', 'attention_mask'] no answer in 200s during put to storage unit su_7c31 at 10.0.0.7:5555; serialized_mb=312.4 tcp=up verdict=request_lost_in_flight (...)
```

Verified: `ruff check` / `ruff format`, `python -m compileall -q transfer_queue tutorial tests`, and 598 passed / 10 skipped (excluding the yuanrong e2e test, which needs the optional `openyuanrong-datasystem`).

**Review by 0oshowero0 — APPROVED** [review]
